servicetalk-encoding-api/src/main/java/io/servicetalk/encoding/api/AbstractZipContentCodec.java

```diff
@@ -204,87 +128,343 @@ public final Buffer decode(final Buffer src, final int offset, final int length,
+                        if (theState == TERMINATED) {
+                            return;
+                        }
+
+                        if (stateUpdater.compareAndSet(DecodingOperator.this,
+                                theState, TERMINATED)) {
+                            // Clean up only if the state was not RESERVED,
+                            // otherwise let the signal in-progress handle the cleaning
+                            if (theState != SIGNAL_IN_PROGRESS) {
+                                inflater.end();
+                            }
+
+                            return;
+                        }
+                    }
+                }
+            });
+
+            subscriber.onSubscribe(this.subscription);
+        }
+
+        @Override
+        public void onNext(@Nullable final Buffer src) {
+            assert subscription != null;
 
+            if (src == null) {
+                return;
+            }
+
+            runSeriallyWithTermHandoff(() -> {
                 // onNext will produce AT-MOST N items (as received)
                 try {
                     if (streamDecoder.isFinished()) {
                         throw new IllegalStateException("Stream encoder previously closed but more input arrived ");
                     }
 
-                    Buffer part = allocator.newBuffer(chunkSize);
+                    Buffer part = allocator.newBuffer(codec.chunkSize);
                     streamDecoder.decode(src, part);
                     if (part.readableBytes() > 0) {
                         subscriber.onNext(part);
+                    } else {
+                        // Not enough data to decompress, ask for more
+                        subscription.request(1);
                     }
-
-                    // Not enough data to decompress, ask for more
-                    subscription.request(1);
                 } catch (Exception e) {
-                    LOGGER.error("Error while decoding with {}", name(), e);
-                    onError(e);
+                    throwException(e);
                 }
-            }
 
-            @Override
-            public void onError(final Throwable t) {
-                assert inflater != null;
+                return AVAILABLE;
+            });
+        }
+
+        @Override
+        public void onError(final Throwable t) {
+            runSeriallyWithTermHandoff(() -> TERMINATED);
+            subscriber.onError(t);
+        }
 
-                inflater.end();
-                subscriber.onError(t);
+        @Override
+        public void onComplete() {
+            runSeriallyWithTermHandoff(() -> TERMINATED);
+            subscriber.onComplete();
+        }
+
+        // Execute a signal while maintaining ability to receive a cancellation handoff that will trigger the clean-up
+        // phase of the resources.
+        private void runSeriallyWithTermHandoff(final IntSupplier work) {
```

Comment:
`DecodingOperator` and `EncodingOperator` seem to be doing the same thing (e.g. manage state that has a lifetime and must be closed), with the exception `EncodingOperator` also needs to insert an `onNext` (which is done in `scanWith`). Can we share this code somehow?